### PR TITLE
feat: allow setting the Origin header and Sec-Fetch-* headers in net.request()

### PR DESCRIPTION
**Pull request by LuoJinghua**
#### Description of Change
This pull request is based on #26134. It's possible to reproduce a request send by Chrome Browser with #26134 and this pull request. 

Is ok to backport this PR to stable branches, such as 10-x-y?

#### Checklist

- [ ] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Allowed setting the 'Origin' header and Sec-Fetch-* headers in net.request().

### CODE REVIEW

File: shell/browser/api/electron_api_url_loader.cc
Comment (nornagon) [inline shell/browser/api/electron_api_url_loader.cc:387]: It looks like we'll start sending `Sec-Fetch-User: ?0` with all requests by default as of this change. Is that desirable?

File: spec-main/api-net-spec.ts
Comment (nornagon) [inline spec-main/api-net-spec.ts:864]:
```suggestion
      it(`should set sec-fetch-mode to ${mode} if requested`, async () => {
```

File: spec-main/api-net-spec.ts
Comment (nornagon) [inline spec-main/api-net-spec.ts:900]:
```suggestion
      it(`should set sec-fetch-dest to ${dest} if requested`, async () => {
```

**Review by nornagon — CHANGES_REQUESTED** [review]
mostly lgtm! just want clarity on the (possibly) new `sec-fetch-user` behavior.

File: docs/api/client-request.md
Comment (nornagon) [inline docs/api/client-request.md:50]: Should we perhaps only allow setting this through the headers?

Comment (LuoJinghua) [reply docs/api/client-request.md:50]: We need to set the origin url but not the origin header for sometime.

File: shell/browser/api/electron_api_url_loader.cc
Comment (LuoJinghua) [reply shell/browser/api/electron_api_url_loader.cc:387]: I don’t think so. But I will add a test to ensure this.

Comment (LuoJinghua) [reply shell/browser/api/electron_api_url_loader.cc:387]: It seems the chromimum network stack will remove the Sec-Fetch-User header if has_user_activation is false.

File: spec-main/api-net-spec.ts
Comment (LuoJinghua) [reply spec-main/api-net-spec.ts]: Done.

File: spec-main/api-net-spec.ts
Comment (LuoJinghua) [reply spec-main/api-net-spec.ts]: Done.

File: shell/browser/api/electron_api_url_loader.cc
Comment (LuoJinghua) [reply shell/browser/api/electron_api_url_loader.cc:387]: I added a test to make sure this doesn't happen.

**Review by zcbenz — APPROVED** [review]

**Comment by zcbenz** [comment]
> Is ok to backport this PR to stable branches, such as 10-x-y?

Once this API is approved by wg-api, we can ask wg-releases for approval of backport.

File: spec-main/api-net-spec.ts
Comment (nornagon) [inline spec-main/api-net-spec.ts:847]: This can be a single test:

```suggestion
    it('should not send sec-fetch-user header by default', async () => {
      const serverUrl = await respondOnce.toSingleURL((request, response) => {
        expect(request.headers).not.to.have.property('sec-fetch-user');
        response.statusCode = 200;
        response.statusMessage = 'OK';
        response.end();
      });
      const urlRequest = net.request({
        url: serverUrl
      });
      await collectStreamBody(await getResponse(urlRequest));
    });
```

**Review by nornagon — APPROVED** [review]
Generally speaking we don't backport new features to stable branches unless there's a specific need. Backporting a new feature to a stable branch requires a minor version bump, and it's up to @electron/wg-releases to decide on a case-by-case basis whether to accept a feature backport.

In my opinion (as a non-member of the Releases WG), this feature presents a non-zero but quite small stability risk. Code paths are changed by this PR, but in such a way that it seems quite unlikely to cause any observable behavior change unless the caller is using the added features.

File: spec-main/api-net-spec.ts
Comment (LuoJinghua) [reply spec-main/api-net-spec.ts]: Done.

**Comment by LuoJinghua** [comment]
Hi, the cors request always fail with net error -2 and cors error kPreflightWildcardOriginNotAllowed. Do you have any idea what happens for this case?

Response of preflight request:
``` 
HTTP/1.1 200 
date	Wed, 28 Oct 2020 00:49:47 GMT
access-control-allow-origin	*
access-control-allow-methods	GET, POST, OPTIONS
access-control-expose-headers	
access-control-max-age	3000
access-control-allow-headers	content-type,x-csrf-token
Content-Security-Policy	frame-ancestors 'self'
strict-transport-security	max-age=31536000; includeSubDomains
Transfer-Encoding	chunked
Connection	keep-alive
```

**Comment by LuoJinghua** [comment]
> Hi, the cors request always fail with net error -2 and cors error kPreflightWildcardOriginNotAllowed. Do you have any idea what happens for this case?
> 
> Response of preflight request:
> 
> ```
> HTTP/1.1 200 
> date	Wed, 28 Oct 2020 00:49:47 GMT
> access-control-allow-origin	*
> access-control-allow-methods	GET, POST, OPTIONS
> access-control-expose-headers	
> access-control-max-age	3000
> access-control-allow-headers	content-type,x-csrf-token
> Content-Security-Policy	frame-ancestors 'self'
> strict-transport-security	max-age=31536000; includeSubDomains
> Transfer-Encoding	chunked
> Connection	keep-alive
> ```

Ok, figured out this finally. We need to backport 0e7d59dd7(#25284) too to fix this issue for stable branches.

File: lib/browser/api/net.ts
Comment (MarshallOfSound) [inline lib/browser/api/net.ts:413]: Headers are case insensitive when read by the networking stack but this assumes that `origin` is lowercase here.  To be accurate this should somehow pull `origin` in a case insensitive way right?

**Comment by jkleinsc** [comment]
The API WG approved this PR at our Nov 9, 2020 meeting

File: lib/browser/api/net.ts
Comment (LuoJinghua) [reply lib/browser/api/net.ts:413]: The header name are stored in lowercase internally. This is another issue we may fix in the future.

```
  setHeader (name: string, value: string) {
    ...
    const key = name.toLowerCase();
    this._urlLoaderOptions.extraHeaders[key] = value;
  }
```

Comment (LuoJinghua) [reply lib/browser/api/net.ts]: I just go ahead and fix this issue too.

File: lib/browser/api/net.ts
Comment (nornagon) [inline lib/browser/api/net.ts:252]: This tuple type is somewhat hard to read, I'd recommend instead using an object:
```suggestion
  const urlLoaderOptions: NodeJS.CreateURLLoaderOptions & { redirectPolicy: RedirectPolicy, headers: Record<string, {originalKey: string, value: string | string[]}> } = {
```

Comment (LuoJinghua) [reply lib/browser/api/net.ts]: Fixed now.

**Pull request merged**
